src/App.tsx

```diff
@@ -1,27 +1,101 @@
+import 'bulma/css/bulma.min.css';
+import { Vortex } from 'react-loader-spinner';
+
 import './App.scss';
-import { GoodsList } from './GoodsList';
+import './reset.scss';
+import './normalize.scss';
+
+import { GoodsList } from './components/GoodList';
+import { Button } from './components/Button';
+
+import { Good } from './types/Good';
+import { ButtonType } from './types/ButtonType';
+
+import { getAll, get5First, getRedGoods } from './api/goods';
+
+export const App: React.FC = () => {
+  const [goods, setGoods] = useState<Good[]>([]);
+  const [clickedButton, setClickedButton] = useState('');
+  const [isLoading, setIsLoading] = useState(false);
+  const [hasError, setHasError] = useState(false);
+
+  const handleClick = useCallback(
+    async (loadGoods: () => Promise<Good[]>, buttonName: string) => {
+      if (clickedButton === buttonName) {
+        return;
+      }
+
+      setIsLoading(true);
+      setHasError(false);
+
+      try {
+        setGoods(await loadGoods());
+        setClickedButton(buttonName);
+      } catch {
+        setHasError(true);
+      }
 
-// import { getAll, get5First, getRed } from './api/goods';
-// or
-// import * as goodsAPI from './api/goods';
+      setIsLoading(false);
+    }, [clickedButton],
+  );
 
-export const App: React.FC = () => (
-  <div className="App">
-    <h1>Dynamic list of Goods</h1>
+  return (
+    <div className="App page page__container">
+      <h1 className="page page__title">
+        Dynamic list of Goods
+      </h1>
 
-    <button type="button" data-cy="all-button">
-      Load all goods
-    </button>
+      <div className="buttons">
+        <Button
+          dataCy="all-button"
+          onClick={() => handleClick(getAll, ButtonType.ALL)}
+          clickedButton={clickedButton}
+        >
+          Load all goods
+        </Button>
 
-    <button type="button" data-cy="first-five-button">
-      Load 5 first goods
-    </button>
+        <Button
+          dataCy="first-five-button"
+          onClick={() => handleClick(get5First, ButtonType.FIVE)}
+          clickedButton={clickedButton}
+        >
+          Load 5 first goods
+        </Button>
 
-    <button type="button" data-cy="red-button">
-      Load red goods
-    </button>
+        <Button
+          dataCy="red-button"
+          onClick={() => handleClick(getRedGoods, ButtonType.RED)}
+          clickedButton={clickedButton}
+        >
+          Load red goods
+        </Button>
+      </div>
 
-    <GoodsList goods={[]} />
-  </div>
-);
+      {isLoading
+        ? (
+          <div>
+            <Vortex
+              height="80"
+              width="80"
+              ariaLabel="vortex-loading"
+              wrapperStyle={{}}
+              wrapperClass="vortex-wrapper"
+              colors={['red', 'green', 'blue', 'yellow', 'orange', 'purple']}
+            />
+          </div>
+        )
+        : (
+          <>
+            {hasError && !isLoading ? (
```

Comment:
```suggestion
            {hasError ? (
```

It's already loaded as you are inside ternary for this case